concepts/dynamic-memory-management/about.md

```diff
@@ -0,0 +1,285 @@
+It's important to remember that, although dynamically allocated memory is claimed by the system until freed, the pointer it is assigned to may only have block scope.
+For instance, in the `malloc` examples, memory is allocated in `my_function`.
+That memory would be leaked if the pointer to it were not returned and used to set the `numbers` pointer in `main`.
+Thus, passing that `numbers` pointer to `free` releases the memory allocated in `my_function`.
+
+#### use after free (**UAF**)
+
+If memory is accessed after it is freed, the value(s) could be anything.
+In the following example, the memory pointed to by `numbers` is freed and then is read from in `printf`.
+
+```c
+#include <stdio.h>
+#include <stdlib.h>
+
+int * my_function(int count) {
+    // pointer-to-void from malloc is converted to a pointer-to-int
+    int * numbers = (int *) malloc(sizeof(int) * count);
+    return numbers;
+}
+
+int main() {
+    int * numbers = my_function(1);
+    if (!numbers) return -1;
+    *numbers = 1;
+    // prints numbers is 1
+    printf("numbers is %d\n", *numbers);
+    free(numbers);
+    // could  print anything, e.g.
+    // The first element of the array is 18178656
+    printf("The first element of the array is %d\n", numbers[0]);
+}
+```
+
+The `numbers` pointer, after it has been freed, is referred to as a dangling pointer.
+A dangling pointer can refer to other memory than that for which it was intended.
+Thus, a dangling pointer can lead to data corruption and arbitrary code substitution and execution.
+A pointer to allocated memory which has not been cleared by setting to `NULL` can be used to hack the program.
+In the above example, if `numbers` was set to `NULL` like so: `numbers = NULL;` after being freed, the following `printf` would result in program termination, perhaps reporting a segmentation fault when attempting to read `numbers[0]`.
+
+#### double free
+
+If `free` is called twice on a pointer to dynamically allocated memory, it results in undefined behavior (**UB**.)
+This means the compiler can arbitrarily handle it any way it wants to.
+The following example, compiled with two different compilers, can have two different outcomes
+
+```c
+#include <stdio.h>
+#include <stdlib.h>
+
+int * my_function(int count) {
+    // pointer-to-void from malloc is converted to a pointer-to-int
+    int * numbers = (int *) malloc(sizeof(int) * count);
+    return numbers;
+}
+
+int main() {
+    int * numbers = my_function(1);
+    if (!numbers) return -1;
+    *numbers = 1;
+    // prints numbers is 1
+    printf("numbers is %d\n", *numbers);
+    free(numbers);
+    // first compiler generates runtime error: double free detected in tcache 2
+    // second compiler generates no error
+    free(numbers);
+}
+```
+
+The first compiler generated no runtime error when `numbers` was set to `NULL` between each call to `free`, but that would not be guaranteed to work with a third compiler.
```

Comment:
I read this as there should be a `numbers = NULL;` line somewhere in the above example, specifically between the two calls to `free()`. But I don't see such. 
Do I misunderstand?

Reply:
Good catch! I just put it in a commented line, since the program was run both with and without setting `numbers = NULL`.